RIOT-OS/RIOT
# ieee802154_submac: add initial support for common MAC sub layer

### Contribution description
This PR adds initial support for the IEEE 802.15.4 SubMAC (#13376). Although this component was introduced in the draft PR #14787 , this PR should contain the "merge-able" version.

In a nutshell, this module provides a uniform layer to send and receive IEEE 802.15.4 MAC frames with retransmissions and CSMA-CA. The SubMAC emulates in software hardware features (CSMA-CA, Retransmissions, etc) missing on a given radio.
This component also keeps track of certain MAC Information Base parameters like channel settings, TX power, etc. An upper layer might (in the near future) request such information directly from the SubMAC layer instead of fetching it from a network device.

Besides that, I added a a `netdev_ieee802154_submac` as a transition layer. With this, default and gnrc_networking are indirectly working on top on the Radio HAL (#13943 ). 

Since the Radio HAL (#14371 ) is asynchronous, #11263 is required in order to send packets with several 6LowPAN fragments.

This module heavily improves the performance of radios without retransmissions / CSMA-CA handling (see e.g https://github.com/RIOT-OS/RIOT/pull/14787#issuecomment-676056009).

In the future, the full IEEE 802.15.4 MAC layer (to be implemented) can benefit of this, as well as applications interested in the lower part of the MAC layer (CSMA-CA, retransmissions, etc) like `lwip`.




### Testing procedure
So far, only the CC2538 has an implementation of the Radio HAL API, but this might change depending on the status of #14792.
- Test that the old `netdev` abstraction still works as expected with GNRC.
- Use the test application in `tests/ieee802154_submac` to send and receive packets using the SubMAC.
- Try `gnrc_networking` and `default` using the `netdev_ieee802154_submac` module. E.g `USEMODULE=netdev_ieee802154_submac BOARD=cc2538dk make -C examples/gnrc_networking all flash term`. Try to run some tests and compare the performance for radios without CSMA-CA or retransmissions. I suggest running the Release tests.

Here are some test results for `samr21-xpro` and `remote-revb` using this PR (both with CSMA-CA and retransmissions):

- Ping payload size: 1024 bytes
- Ping interval: 170 ms
- Count: 1000

```
2020-09-04 15:19:16,076 # --- fe80::2068:3123:fe42:2e25 PING statistics ---
2020-09-04 15:19:16,083 # 1000 packets transmitted, 1000 packets received, 0% packet loss
2020-09-04 15:19:16,088 # round-trip min/avg/max = 133.318/145.909/161.423 ms
```

Disabling retransmissions in `cc2538` (similar to what we have in master):
```
020-09-04 15:29:11,891 # --- fe80::2068:3123:fe42:2e25 PING statistics ---
2020-09-04 15:29:11,897 # 1000 packets transmitted, 626 packets received, 37% packet loss
2020-09-04 15:29:11,901 # round-trip min/avg/max = 129.693/140.842/155.274 ms
```




### Issues/PRs references
#13376 
#14787 
#11263
